handlers/event-handler.ts

```diff
@@ -1 +1,158 @@
-export {};
+import { EventPattern } from 'aws-cdk-lib/aws-events';
+import type {
+	EventBridgeEvent,
+	SQSHandler,
+	SQSMessageAttribute,
+	SQSRecord,
+} from 'aws-lambda';
+import { AsyncHandler, HandlerDefinition, HandlerTypes } from './handler';
+import type { InvalidMessage, Message, ValidatedMessage } from './message';
+
+export interface EventHandlerDefinition extends HandlerDefinition {
+	/**
+	 * Event handler name
+	 */
+	name: string;
+
+	/**
+	 * Uses event bus name that is provided and configured within stack.
+	 * If no name is specified, the first bus listed in the CDK configuration is used.
+	 */
+	eventBusName?: string;
+
+	/**
+	 * The event pattern in which to subscribe to.
+	 */
+	eventPattern: EventPattern;
+
+	/**
+	 * The largest number of records that AWS Lambda will retrieve from your event source at the time of invoking your function.
+	 *
+	 * Valid Range: Minimum value of 1. Maximum value of 10.
+	 * If `maxBatchingWindow` is configured, this value can go up to 10,000.
+	 *
+	 * @default 10
+	 */
+	batchSize?: number;
+
+	/**
+	 * The maximum amount of time to gather records before invoking the function.
+	 *
+	 * Valid Range: Minimum value of 0 minutes. Maximum value of 5 minutes.
+	 *
+	 * @default - no batching window. The lambda function will be invoked immediately with the records that are available.
+	 */
+	maxBatchingWindow?: number;
+
+	/**
+	 * How many times to attempt processing a message
+	 *
+	 * @default 10
+	 */
+	maximumAttempts?: number;
+}
+
+export type EventHandlerEvent<T> = {
+	/**
+	 * The raw SQS records
+	 */
+	Records: SQSRecord[];
+	/**
+	 * Messages that have been validated
+	 */
+	ValidMessages: ValidatedMessage<EventBridgeEvent<string, T>>[];
+	/**
+	 * Messages that failed to validate
+	 */
+	InvalidMessages: InvalidMessage[];
+};
+
+export interface EventHandlerOptions<T> extends EventHandlerDefinition {
+	validator?: (
+		messageBody: EventBridgeEvent<string, any>,
+	) => EventBridgeEvent<string, T>;
```

Comment:
This doesn't seem like it's a "validator" per-say since it actually takes AND returns the event.

I'd expect the signature of this to be something more like:

```typescript
export interface EventHandlerOptions<TBody> extends EventHandlerDefinition {
	validator?: (
		messageBody: EventBridgeEvent<string, unknown>,
	) => messageBody is EventBridgeEvent<string, TBody>;
}
```

Reply:
If you change it to this, you need to also change the usage of validator to be:
```typescript
                                 const parsedBody = JSON.parse(record.body);
				if (validator && validator(parsedBody)) {
					queueEvent.ValidMessages.push({
						body: parsedBody,
						messageId: record.messageId,
						attempts: attempts,
					});
				}
```

Reply:
I like this, but the way it is now is how it is within the entire faceteer ecosystem. I want to use your pattern though for when I tackle the issue https://github.com/faceteer/cdk/issues/20.

Reply:
I've decided using this pattern has limitations. Some validators might want to return the data differently as it comes in. For example, Joi has a default setting which will automatically convert a string into a number if necessary.